patches/server/0969-New-event-verbose-logging-command.patch

```diff
@@ -0,0 +1,204 @@
++                isRunning = false;
++            }
++        }else {
++            isRunning = false;
++            sender.sendMessage(text("Logging events..", GREEN));
++            try {
++                logWrappers();
++                sender.sendMessage(text("Logged all event verbose to " + currentFile, GREEN));
++                currentFile = null;
++                toDump.clear();
++            }catch(IOException e) {
++                sender.sendMessage(text("Failed to create event verbose files! See the console for more info.", RED));
++                MinecraftServer.LOGGER.warn("Error occurred while dumping event verbose", e);
++            }
++        }
++        return true;
++    }
++
++    /**
++     * Dumps all currently stored {@link Wrapper}'s to our {@link #currentFile}
++     * @throws IOException If something broke writing it
++     */
++    private void logWrappers() throws IOException {
++        JsonArray array = new JsonArray();
++        toDump.forEach(w -> {
++            @Nullable JsonObject obj = w.serialize();
++            if(obj != null) //will only be null if an event class only has final fields
++                array.add(obj);
++        });
++
++        StringWriter stringWriter = new StringWriter();
++        JsonWriter jsonWriter = new JsonWriter(stringWriter);
++        jsonWriter.setIndent(" ");
++        jsonWriter.setLenient(false);
++        Streams.write(array, jsonWriter);
++
++        try (PrintStream out = new PrintStream(Files.newOutputStream(currentFile), false, StandardCharsets.UTF_8)) {
++            out.print(stringWriter);
++        }
++
++    }
++
++    /**
++     * Called in {@link io.papermc.paper.plugin.manager.PaperPluginManagerImpl#callEvent(Event)} to save in {@link #toDump}
++     */
++    public static void tryLogEvent(Event ev) {
++        if(!isRunning)
++            return;
++        if(exclude.contains(ev.getClass()))
++            return;
++        toDump.add(new Wrapper(ev));
++    }
++
++    /**
++     * Wrapper class for QoL Serialization, just stores the event object and the time it ran to log eventually
++     */
++    private static final class Wrapper {
++        private final String timestamp;
++        private final Event event;
++
++        public Wrapper(Event event) {
++            this.timestamp = FORMATTER.format(LocalDateTime.now());
++            this.event = event;
++        }
++
++        @Nullable
++        /**
++         * Properly serialize our Wrapper into a JsonObject, so we can log it properly
++         * @implNote Is null if the event has only final fields
++         */
++        private JsonObject serialize() {
++            JsonObject parent = new JsonObject();
++            parent.addProperty("timestamp", timestamp);
++            JsonObject obj = new JsonObject();
++            obj.addProperty("event", event.getEventName());
++            try {
++                Field[] fields = event.getClass().getDeclaredFields();
```

Comment:
This will ignore important fields like the Block or Entity in question of BlockEvent or EntityEvent since the field for that is stored on the superclass. You need to collect important fields from all the types in the hierarchy.